你好，我是孔令飞。

在Go项目开发中，错误是我们必须要处理的一个事项。除了我们上一讲学习过的错误码，处理错误也离不开错误包。

业界有很多优秀的、开源的错误包可供选择，例如Go标准库自带的`errors`包、`github.com/pkg/errors`包。但是这些包目前还不支持业务错误码，很难满足生产级应用的需求。所以，在实际开发中，我们有必要开发出适合自己错误码设计的错误包。当然，我们也没必要自己从0开发，可以基于一些优秀的包来进行二次封装。

这一讲里，我们就来一起看看，如何设计一个错误包来适配上一讲我们设计的错误码，以及一个错误码的具体实现。

## 错误包需要具有哪些功能？

要想设计一个优秀的错误包，我们首先得知道一个优秀的错误包需要具备哪些功能。在我看来，至少需要有下面这六个功能：

**首先，应该能支持错误堆栈。**我们来看下面一段代码，假设保存在[bad.go](https://github.com/marmotedu/gopractise-demo/blob/master/errors/bad.go)文件中：

```
package main

import (
	"fmt"
	"log"
)

func main() {
	if err := funcA(); err != nil {
		log.Fatalf("call func got failed: %v", err)
		return
	}

	log.Println("call func success")
}

func funcA() error {
	if err := funcB(); err != nil {
		return err
	}

	return fmt.Errorf("func called error")
}

func funcB() error {
	return fmt.Errorf("func called error")
}
```

执行上面的代码：

```
$ go run bad.go
2021/07/02 08:06:55 call func got failed: func called error
exit status 1
```

这时我们想定位问题，但不知道具体是哪行代码报的错误，只能靠猜，还不一定能猜到。为了解决这个问题，我们可以加一些Debug信息，来协助我们定位问题。这样做在测试环境是没问题的，但是在线上环境，一方面修改、发布都比较麻烦，另一方面问题可能比较难重现。这时候我们会想，要是能打印错误的堆栈就好了。例如：
<div><strong>精选留言（30）</strong></div><ul>
<li><img src="https://static001.geekbang.org/account/avatar/00/0f/87/64/3882d90d.jpg" width="30px"><span>yandongxiao</span> 👍（10） 💬（1）<div>总结：
1. 为什么业务要开发自己的错误包？因为开源的error包中，没有业务错误码，不方便使用。
2. 如何区分内部错误和外部错误？
    1. 将错误分为两部分，是一个非常好的设计，通过 error code 联系在一起
    2. 将 error code, response code, external error message 注册到一张大表上
    3. github.com&#47;marmotedu&#47;errors 包提供的 WitheCode 和 Wrap 方法，构建错误链（嵌套关系）
    4. 通过 format 的各种形式，控制 error 的输出信息
3. 如何自动生成全局的错误码文档，方便排查问题和联调？ 使用 codegen 工具。
4. 如何打印错误？只在错误最原始位置打印错误。</div>2021-11-27</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/12/52/40/e57a736e.jpg" width="30px"><span>pedro</span> 👍（9） 💬（1）<div>以推荐老师的项目给组里其它小伙伴，一起学习沉淀打磨为自己的实用标准，多谢老师</div>2021-07-08</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/12/a7/b2/274a4192.jpg" width="30px"><span>漂泊的小飘</span> 👍（5） 💬（1）<div>感觉需要看着源码再学一遍。。。</div>2021-07-28</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/25/85/17/91ad99c3.jpg" width="30px"><span>树建</span> 👍（3） 💬（3）<div>文中提到withCode实现了format方法，但是看了errors.go中只查看withMessage实现了format</div>2022-05-05</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/18/8a/ac/a44b4b3f.jpg" width="30px"><span>吹口哨yu</span> 👍（2） 💬（1）<div>withCode 没有实现 Format() 方法呀？</div>2022-07-13</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/2b/ed/da/78a0e68f.jpg" width="30px"><span>.</span> 👍（2） 💬（1）<div>看完了git中withCode错误的实现逻辑对于 WitCode需要分开cause和err 两个errors，很是迷惑。起初看完代码在猜想cause这个应该是为了递归打印，实现上个章节的”科学设计错误包“内容“ 错误码100101， 10: 服务。01: 某个服务下的某个模块。01: 模块下的错误码序号，每个模块可以注册 100 个错误。” 的功能，但是看了函数IsCode的逻辑，彻底懵逼了，code只对比了第一层的code就return，显然不是为了实现上一章节的内容，所以多个cause  errors感觉是一个特别累赘的设计。

另外WithCode和WithStack 的入参都是error 接口。 如果一个error起初是go原生库的error， 先withCode再WithMessgge, 在执行err. Error() 会丢失打印不出完整的错误栈。这块是原本iam的就是业务设计就是如此吗？WithMessgge会直接决定了err. Error() 值，不管前面被With了多少次？</div>2022-06-11</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/25/06/13/e4f9f79b.jpg" width="30px"><span>你赖东东不错嘛</span> 👍（2） 💬（3）<div>记录错误方式二：在err发生处打印了log，却依旧把err上抛，而最外层又对err进行了一次处理，这样可能导致日志里写了两份重复的err信息。望解惑！</div>2021-07-16</li><br/><li><img src="https://thirdwx.qlogo.cn/mmopen/vi_32/5JKZO1Ziax3Ky03noshpVNyEvZw0pUwjLcHrHRo1XNPKXdmCE88homb6ltA15CdVRnjzjgGs3Ex42CaDbeYzNuQ/132" width="30px"><span>Geek_25f93f</span> 👍（1） 💬（1）<div>要写这么一个codegen工具应该要废很多时间</div>2022-07-09</li><br/><li><img src="http://thirdwx.qlogo.cn/mmopen/vi_32/4Lprf2mIWpJOPibgibbFCicMtp5bpIibyLFOnyOhnBGbusrLZC0frG0FGWqdcdCkcKunKxqiaOHvXbCFE7zKJ8TmvIA/132" width="30px"><span>Geek_c2089d</span> 👍（1） 💬（1）<div>func New(message string) error {
	return &amp;fundamental{
		msg:   message,
		stack: callers(),
	}
}
老师，想问一下errors&#47;code.go 的包里面的callers()函数，哪里来的</div>2022-04-19</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/10/95/00/a7001873.jpg" width="30px"><span>莫林</span> 👍（1） 💬（1）<div>老师能介绍一下如何适配gprc 的 status 响应吗？</div>2021-12-07</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/12/de/97/cda3f551.jpg" width="30px"><span>倪昊</span> 👍（1） 💬（1）<div>错误码实际使用方法实例里的WriteResponse，如果err != nil，if分支执行完，还是会继续执行下面的c.JSON，这样不就返回两个Response结构了吗？</div>2021-09-08</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/11/7a/d2/4ba67c0c.jpg" width="30px"><span>Sch0ng</span> 👍（1） 💬（1）<div>把错误处理落地成简单实用的错误处理包，有助于错误码的长期治理。</div>2021-08-10</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/15/27/2d/120c21ef.jpg" width="30px"><span>大刘</span> 👍（1） 💬（1）<div>错误码实际使用方法示例中
1.type Response struct { Code errors.Code `json:&quot;code,omitempty&quot;` Message string `json:&quot;message,omitempty&quot;` Reference string `json:&quot;reference,omitempty&quot;` Data interface{} `json:&quot;data,omitempty&quot;`}

Code errors.Code 是不是应该改为Code int
2.if err != nil { core.WriteResponse(c, code.ErrUserNotFound.Error(), nil) return }
code.ErrUserNotFound.Error()  这个Error方法是从哪里来的？ ErrUserNotFound按照上面所说应该只是一个int值吧？</div>2021-08-02</li><br/><li><img src="https://thirdwx.qlogo.cn/mmopen/vi_32/Q0j4TwGTfTKotsBr2icbYNYlRSlicGUD1H7lulSTQUAiclsEz9gnG5kCW9qeDwdYtlRMXic3V6sj9UrfKLPJnQojag/132" width="30px"><span>ppd0705</span> 👍（1） 💬（2）<div>请问一下老师，当一个接口有多个用户输入参数，其中某个参数错误, 如果需要告诉用户具体哪个参数有问题，这样CODE.Ext是动态的，该怎么处理好呢？ 


</div>2021-07-30</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/16/60/4f/db0e62b3.jpg" width="30px"><span>Daiver</span> 👍（1） 💬（1）<div>还是不清楚，codegen是怎么生成注册代码的</div>2021-07-12</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/15/2e/01/14a478bb.jpg" width="30px"><span>happychap</span> 👍（1） 💬（2）<div>老师，您好，学习了您的专栏受益良多，谢谢您的付出！有个问题还不太明白，烦请解答一下，就是：接口请问返回给error处理ref链接内容来源是怎么形成的或该如何维护管理它们呢？</div>2021-07-08</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/2c/2b/8e/4d24c872.jpg" width="30px"><span>season</span> 👍（0） 💬（1）<div>如何记录错误 章节中的两种记录错误的方式更推荐哪个呢？
或者说两种方式各自的使用场景是怎样的？改如何进行选择？</div>2022-09-27</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/12/d6/98/e2d8f2a9.jpg" width="30px"><span>zzq</span> 👍（0） 💬（1）<div>偶偶、 老师、 刚才提交的留言是关于 注释首字母必须大写的问题、 我自己编译一下把正则 首字母大写去掉了。 

学习了很多， 大佬牛批～～～～。</div>2022-09-26</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/12/d6/98/e2d8f2a9.jpg" width="30px"><span>zzq</span> 👍（0） 💬（1）<div>老师、我在单独练习 codegen -type=int 生成错误码时、发现

&#47;&#47; ErrSuccess - 200: 成功.
ErrSuccess int = iota + 100001

生成失败 是因为 成功这个要求首字母大写， 我想使用纯汉字的描述 应该怎么做。 </div>2022-09-26</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/1f/f3/ed/46299341.jpg" width="30px"><span>打工人yyds</span> 👍（0） 💬（1）<div>老师，我看您的源码，apiservere的gin-jwt认证那块的返回(Unauthorized)并没有使用统一错误格式。请问这块一般不设置吗？</div>2022-03-23</li><br/><li><img src="https://thirdwx.qlogo.cn/mmopen/vi_32/Q0j4TwGTfTLDUJyeq54fiaXAgF62tNeocO3lHsKT4mygEcNoZLnibg6ONKicMgCgUHSfgW8hrMUXlwpNSzR8MHZwg/132" width="30px"><span>types</span> 👍（0） 💬（1）<div>github.com&#47;marmotedu&#47;errors是基于https:&#47;&#47;github.com&#47;pkg&#47;errors进行的修改
1.感觉代码风格跟pkg&#47;errors不是很一致
2. error包需要支持code，这个很好，但是不清楚为什么还要保留原理的withMessage, withStack, 然后花大量代码来处理， 只保留withCode相关的是否会更好，更加简洁？？ </div>2022-02-16</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/1f/26/34/891dd45b.jpg" width="30px"><span>宙斯</span> 👍（0） 💬（1）<div>为什么我们在讲解错误码实现时，只讲了withCode没有讲withMessage，并且文中提到withCode有formater，但实际上只有withMessage有formater函数呢？</div>2021-11-07</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/12/37/92/961ba560.jpg" width="30px"><span>授人以🐟，不如授人以渔</span> 👍（0） 💬（1）<div>WriteResponse 函数定义中，如果 err != nil 的情况下，调用 c.JSON 后是需要 return 的。</div>2021-10-14</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/21/5f/f8/1d16434b.jpg" width="30px"><span>陈麒文</span> 👍（0） 💬（1）<div>直接拿老师的这个errors包，用起来</div>2021-08-18</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/28/83/17/df99b53d.jpg" width="30px"><span>随风而过</span> 👍（3） 💬（0）<div>错误码的设计很合理，和其他语言的错误设计大相径庭，很规范，老师这套错误可以引入到企业开发项目当中</div>2021-09-01</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/17/bf/86/c0cb35f0.jpg" width="30px"><span>8.13.3.27.30</span> 👍（1） 💬（0）<div>非常实用的东西、老师讲的东西在实际应用中、真真切切的用到了、可惜当时老师没有出这个教程、现在只能下次重构的时候再使用了</div>2021-07-12</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/1d/04/6a/6aa6c0ea.jpg" width="30px"><span>beshall</span> 👍（0） 💬（0）<div>老师，我跑这个 codegen 老是会  panic 呢？ 而且报错的地方还很深，定位不出来</div>2024-06-13</li><br/><li><img src="" width="30px"><span>Geek_7a540f</span> 👍（0） 💬（0）<div>我在项目中使用的go1.22版本，也顺利启动了服务，但是在执行错误码生成的操作时报错(导入的包没有任何导出类型),清空了缓存也检查了导入的包和路径以及依赖均没问题。最后将go版本降到1.18.3,修改work和mod的go版本,清空缓存并下载依赖，重新安装codegen,最后成功生成代码</div>2024-06-12</li><br/><li><img src="https://static001.geekbang.org/account/avatar/00/36/85/6e/1d1caacd.jpg" width="30px"><span>青云</span> 👍（0） 💬（0）<div>还是需要动手实践一遍才能理解</div>2023-08-29</li><br/><li><img src="https://thirdwx.qlogo.cn/mmopen/vi_32/O9jhHevKia7Axa7dAiaFCmLIFZT3uaE47R09Mkxk54Fw7dcv7yS4yFFW8JlvxEzLJY3AUcKdsyeViap8xIDJzdj1w/132" width="30px"><span>Geek_9b19ad</span> 👍（0） 💬（0）<div>日志包好像没有做日志切割,日志包 跟错误包 以及code  三者的结合 非常好使</div>2023-07-18</li><br/>
</ul>